packages/assets-controllers/src/CurrencyRateController.ts

```diff
@@ -215,7 +223,104 @@ export class CurrencyRateController extends StaticIntervalPollingController<Curr
       console.error('Failed to fetch exchange rates.', error);
     }
 
-    // fallback to crypto compare
+    // fallback using spot price from token prices service
+    try {
+      // Step 1: Get all network configurations to find matching chainIds for native currencies
+      const networkControllerState = this.messenger.call(
+        'NetworkController:getState',
+      );
+      const networkConfigurations =
+        networkControllerState.networkConfigurationsByChainId;
+
+      // Step 2: Build a map of nativeCurrency -> chainId(s)
+      const currencyToChainIds = Object.entries(nativeCurrenciesToFetch).reduce(
+        (acc, [nativeCurrency, fetchedCurrency]) => {
+          // Find chainIds that have this native currency
+          const chainIds = (
+            Object.entries(networkConfigurations) as [
+              Hex,
+              NetworkConfiguration,
+            ][]
+          )
+            .filter(
+              ([, config]) =>
+                config.nativeCurrency.toUpperCase() ===
+                fetchedCurrency.toUpperCase(),
+            )
+            .map(([chainId]) => chainId);
+
+          if (chainIds.length > 0) {
+            acc[nativeCurrency] = {
+              fetchedCurrency,
+              chainId: chainIds[0], // Use the first matching chainId
+            };
+          }
+
+          return acc;
+        },
+        {} as Record<string, { fetchedCurrency: string; chainId: Hex }>,
+      );
+
+      // Step 3: Fetch token prices for each chainId
+      const ratesFromTokenPrices = await Promise.all(
+        Object.entries(currencyToChainIds).map(
+          async ([nativeCurrency, { chainId }]) => {
```

Comment:
Can we change this to a batch? Worried of calling `Promise.all` on many chains at once.